feat: preserve file name, support sharing multiple files

lidel opened the pull request:
This PR implements #349  by switching to the latest `js-ipfs-api` which enables support for `wrapWithDirectory` option in `ipfs.files.add`.

## UI Preview

I did not want to do a redesign so just added options at the bottom:
![screenshot_50](https://user-images.githubusercontent.com/157609/40127069-e39c1f3a-592e-11e8-8c76-94de74a7c9e7.png)

Not sure if these labels are clear enough, any suggestions?

## TODO before merge

**Add**

- [x] support sharing multiple files
- [x] add checkbox for controlling directory wrapping 
- [x] add checkbox for controlling automatic pinning (added while we are at it)
- [x] add locale keys

**Fix**

- [x]  disable controls when embedded node is used
- [x] solve `FILE_TOO_LARGE` error from `mozilla/addons-linter` caused by the size of the latest `js-ipfs-api`

**Optional (can be a separate PR)**

- [ ] wait for https://github.com/ipfs/js-ipfs/pull/1329 to be released (js-ipfs > v0.28.2) and switch to it, check fi wrapping works and unlock it for embedded node

lidel commented:
@alanshaw @olizilla  UI is temporary, but is there anything that could be improved before we merge it?

olizilla commented:
"Preserve the filename when adding a single item" is good wording for the label, but it leads to the question, when would I not want that? I vote for just doing the right thing here (wrap) and not troubling the user with the decision, as per https://github.com/ipfs-shipyard/pm-ipfs-gui/issues/15#issuecomment-373936917 I don't think wrapping by default would harm anyone here.

"Recursively Pin added data"... I think we just do this for all data the user adds. I don't see any harm in doing this and not asking the user.

lidel commented:
@olizilla Yeah it may be just me :upside_down_face:, but I  _often_ share ephemeral images and text snippets that don't need a name (or have an ugly one that  I'd have to change before sharing) nor permanence (I don't' care to keep it alive forever). 

I agree, the defaults are sane,  but I really want to give power user an option to opt-out without need of using commandline, so I just collapsed them under a tiny `upload options »` that reveals them on click:

![screenshot_52](https://user-images.githubusercontent.com/157609/40186284-00e7076e-59f5-11e8-83d6-f32de86a2c95.png)

olizilla commented:
"Upload options" is good and clear.

olizilla commented:
I'll give this a try now

olizilla commented:
Using unicode glyphs as icons always seems like a good idea, but the cross browser rendering issues will drive you mad. I have tried.

**extra lefty ticks on FF on mac**
<img width="455" alt="screenshot 2018-05-18 11 06 17" src="https://user-images.githubusercontent.com/58871/40229339-db4ce66c-5a8b-11e8-82a3-43c8621b70b9.png">

**no labels on Chrome on mac**
<img width="416" alt="screenshot 2018-05-18 11 09 53" src="https://user-images.githubusercontent.com/58871/40229402-0697963c-5a8c-11e8-94ef-caaf8dad5faf.png">

olizilla commented:
@lidel i'll see if I can track down why the labels are missing on chrome.

olizilla commented:
@lidel ah, the chrome label issue went away after a rebuild. meh.

Now the labels are hidden by default, I'm thinking they can be more informative like

- [x] `--wrap-with-directory` _Wrap files in a directory to preserve their filenames._
- [x] `--pin` _Pin files so they are retained when performing garbage collection on your local IPFS repo._
> See: https://ipfs.io/docs/commands/#ipfs-add for more info

lidel commented:
Switched to pure CSS checkmark that should look the same everywhere, updated labels.
I don't want to focus on it too much as the screen will be redesigned anyway, 
but it should look ok for now:
![screenshot_55](https://user-images.githubusercontent.com/157609/40235625-24f322a8-5aab-11e8-862e-1faa63e38c14.png)

@olizilla please check if it does not :fire:  on mac.
I'd like to merge this and release to beta ASAP  :)

olizilla commented on add-on/src/lib/ipfs-companion.js:
```diff
@@ -274,19 +276,25 @@ module.exports = async function init () {
-          'url': url
-        })
-        console.info('[ipfs-companion] successfully stored', path)
+        console.log('uploadResultHandler.file-' + result.indexOf(file), file)
```
Remove debug logging or add `[ipfs-companion]` prefix

olizilla commented on add-on/src/lib/ipfs-companion.js:
```diff
@@ -274,19 +276,25 @@ module.exports = async function init () {
+        console.log('uploadResultHandler.file-' + result.indexOf(file), file)
+        // open the wrapping directory (or the CID if wrapping was disabled)
+        if (result.length === 1 || file.path === '' || file.path === file.hash) {
+          await browser.tabs.create({
```
What's the expected behaviour here? If i add a single file and let it be wrapped, I'm seeing it open the directory not the file. If i add multiple, it shows the directory.

lidel replied:
Yes, that is correct.

If wrapping is enabled or you are uploading more than one file, it will **always open wrapping directory** (the default).
If wrapping is disabled and a single file is uploaded, it will be opened directly. (requires explicit opt-in from user via upload options).
 
I feel it is a good default for now:
- Most of users want to keep file names and as a side effect will get a branded directory listing provided by gateway
  (Until we integrate ideas from https://github.com/tableflip/share-via-ipfs via separate PRs)
- Uploaded file can be a `.zip` file and browser can't show that. 
  (Opening the file itself will just close the tab and start the download, which makes it difficult for user to copy the URL)
- Mobile users are not forced to download the data, they just get directory listing and decide what do do with it (open the specific file, copy link to directory, copy direct link to file etc)

olizilla commented on add-on/src/popup/quick-upload.js:
```diff
@@ -62,14 +79,41 @@ function quickUploadStore (state, emitter) {
+    `
+  }
+  return html`
+    <button class='f6 link dim bn bg-transparent moon-gray dib pa0 pointer ma0' style='color: #6ACAD1' onclick=${onExpandOptions}>
```
Add an `mt3` here to get the same vertical space when the button transitions to the options.
